File: spine_visualization.py
import serial
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
import threading
import time
from collections import deque
import math

class SpineVisualizer:
    def __init__(self, com_port=14, baud_rate=115200, max_points=100):
        # 串口设置
        self.serial_port = serial.Serial(f'COM{com_port}', baud_rate, timeout=0.1)
        self.running = False
        
        # 数据存储
        self.max_points = max_points
        self.time_data = deque(maxlen=max_points)
        self.imu_data = deque(maxlen=max_points)
        
        # 脊柱参数
        self.spine_length = 0.4  # 脊柱总长度 (米)
        self.segment_length = self.spine_length / 4  # 每段长度
        
        # 创建图形
        self.fig = plt.figure(figsize=(12, 8))
        self.ax = self.fig.add_subplot(111, projection='3d')
        
        # 设置图形属性
        self.ax.set_xlabel('X (前)')
        self.ax.set_ylabel('Y (左)')
        self.ax.set_zlabel('Z (上)')
        self.ax.set_title('实时脊柱姿态可视化')
        
        # 设置坐标轴范围
        self.ax.set_xlim([-0.3, 0.3])
        self.ax.set_ylim([-0.3, 0.3])
        self.ax.set_zlim([0, 0.5])
        
        # 线程锁
        self.lock = threading.Lock()
        
    def euler_to_rotation_matrix(self, roll, pitch, yaw):
        """欧拉角转旋转矩阵"""
        # 转换为弧度
        roll = math.radians(roll)
        pitch = math.radians(pitch)
        yaw = math.radians(yaw)
        
        # 旋转矩阵
        Rx = np.array([[1, 0, 0],
                      [0, math.cos(roll), -math.sin(roll)],
                      [0, math.sin(roll), math.cos(roll)]])
        
        Ry = np.array([[math.cos(pitch), 0, math.sin(pitch)],
                      [0, 1, 0],
                      [-math.sin(pitch), 0, math.cos(pitch)]])
        
        Rz = np.array([[math.cos(yaw), -math.sin(yaw), 0],
                      [math.sin(yaw), math.cos(yaw), 0],
                      [0, 0, 1]])
        
        return Rz @ Ry @ Rx
    
    def calculate_spine_points(self, imu_angles):
        """根据IMU角度计算脊柱各点位置"""
        points = []
        current_pos = np.array([0.0, 0.0, 0.0])
        current_orientation = np.eye(3)
        
        # 计算每个脊柱段的位置
        for i in range(5):
            # 获取当前IMU的欧拉角
            roll, pitch, yaw = imu_angles[i*3:(i+1)*3]
            
            # 计算旋转矩阵
            R = self.euler_to_rotation_matrix(roll, pitch, yaw)
            
            # 更新当前方向
            current_orientation = R @ current_orientation
            
            # 计算下一段的方向向量 (Z轴方向)
            direction = current_orientation @ np.array([0, 0, 1])
            
            # 计算下一段的位置
            next_pos = current_pos + direction * self.segment_length
            
            points.append(current_pos.copy())
            current_pos = next_pos
        
        # 添加最后一个点
        points.append(current_pos)
        
        return np.array(points)
    
    def read_serial_data(self):
        """读取串口数据"""
        while self.running:
            try:
                if self.serial_port.in_waiting > 0:
                    line = self.serial_port.readline().decode('utf-8').strip()
                    if line:
                        # 解析数据
                        data = line.split(',')
                        if len(data) == 16:  # 时间戳 + 5个IMU * 3个角度
                            timestamp = float(data[0])
                            imu_angles = [float(x) for x in data[1:]]
                            
                            with self.lock:
                                self.time_data.append(timestamp)
                                self.imu_data.append(imu_angles)
                                
            except Exception as e:
                print(f"串口读取错误: {e}")
                time.sleep(0.01)
    
    def update_plot(self):
        """更新图形"""
        while self.running:
            try:
                with self.lock:
                    if len(self.imu_data) > 0:
                        # 获取最新的IMU数据
                        latest_angles = self.imu_data[-1]
                        
                        # 计算脊柱点
                        spine_points = self.calculate_spine_points(latest_angles)
                        
                        # 清除之前的图形
                        self.ax.clear()
                        
                        # 设置坐标轴
                        self.ax.set_xlabel('X (前)')
                        self.ax.set_ylabel('Y (左)')
                        self.ax.set_zlabel('Z (上)')
                        self.ax.set_title('实时脊柱姿态可视化')
                        self.ax.set_xlim([-0.3, 0.3])
                        self.ax.set_ylim([-0.3, 0.3])
                        self.ax.set_zlim([0, 0.5])
                        
                        # 绘制脊柱
                        x, y, z = spine_points.T
                        self.ax.plot(x, y, z, 'b-o', linewidth=3, markersize=8, label='脊柱')
                        
                        # 绘制IMU位置
                        for i, point in enumerate(spine_points[:-1]):
                            self.ax.text(point[0], point[1], point[2], f'IMU{i+1}', 
                                       fontsize=8, color='red')
                        
                        # 添加图例
                        self.ax.legend()
                        
                        # 更新显示
                        plt.pause(0.01)
                        
            except Exception as e:
                print(f"图形更新错误: {e}")
                time.sleep(0.01)
    
    def start(self):
        """启动可视化"""
        self.running = True
        
        # 启动数据读取线程
        self.read_thread = threading.Thread(target=self.read_serial_data)
        self.read_thread.daemon = True
        self.read_thread.start()
        
        # 启动图形更新线程
        self.plot_thread = threading.Thread(target=self.update_plot)
        self.plot_thread.daemon = True
        self.plot_thread.start()
        
        print("脊柱可视化已启动，按 Ctrl+C 停止...")
        
        try:
            # 主循环
            while self.running:
                time.sleep(0.1)
        except KeyboardInterrupt:
            self.stop()
    
    def stop(self):
        """停止可视化"""
        self.running = False
        if hasattr(self, 'serial_port'):
            self.serial_port.close()
        plt.close('all')
        print("可视化已停止")

if __name__ == "__main__":
    # 创建可视化器
    visualizer = SpineVisualizer(com_port=14, baud_rate=115200)
    
    # 启动可视化
    visualizer.start()

File: spine_visualization.m
function spine_visualization()
    % 实时脊柱姿态可视化 - MATLAB版本
    % 串口设置
    com_port = 14;
    baud_rate = 115200;
    
    try
        % 创建串口对象
        s = serial(sprintf('COM%d', com_port), 'BaudRate', baud_rate, 'Timeout', 0.1);
        fopen(s);
        fprintf('成功连接到COM%d\n', com_port);
    catch ME
        fprintf('串口连接失败: %s\n', ME.message);
        return;
    end
    
    % 创建图形窗口
    figure('Name', '实时脊柱姿态可视化', 'NumberTitle', 'off', 'Position', [100, 100, 1200, 800]);
    ax = axes('Parent', gcf);
    hold(ax, 'on');
    grid(ax, 'on');
    
    % 设置坐标轴
    xlabel(ax, 'X (前)');
    ylabel(ax, 'Y (左)');
    zlabel(ax, 'Z (上)');
    title(ax, '实时脊柱姿态可视化');
    axis(ax, [-0.3, 0.3, -0.3, 0.3, 0, 0.5]);
    view(ax, 3);
    
    % 脊柱参数
    spine_length = 0.4;  % 脊柱总长度 (米)
    segment_length = spine_length / 4;  % 每段长度
    
    % 初始化绘图对象
    spine_line = plot3(ax, NaN, NaN, NaN, 'b-o', 'LineWidth', 3, 'MarkerSize', 8);
    imu_points = zeros(5, 1);
    imu_labels = zeros(5, 1);
    
    for i = 1:5
        imu_points(i) = plot3(ax, NaN, NaN, NaN, 'ro', 'MarkerSize', 10);
        imu_labels(i) = text(ax, NaN, NaN, NaN, sprintf('IMU%d', i), 'FontSize', 8, 'Color', 'red');
    end
    
    legend(ax, '脊柱', 'Location', 'best');
    
    % 主循环
    running = true;
    fprintf('启动实时可视化，按 Ctrl+C 停止...\n');
    
    while running
        try
            % 检查串口是否有数据
            if s.BytesAvailable > 0
                % 读取一行数据
                line = fgetl(s);
                
                if ~isempty(line)
                    % 解析数据
                    data = str2double(strsplit(line, ','));
                    
                    if length(data) == 16  % 时间戳 + 5个IMU * 3个角度
                        imu_angles = data(2:end);  % 跳过时间戳
                        
                        % 计算脊柱点
                        spine_points = calculate_spine_points(imu_angles, segment_length);
                        
                        % 更新脊柱线条
                        set(spine_line, 'XData', spine_points(:, 1), ...
                                      'YData', spine_points(:, 2), ...
                                      'ZData', spine_points(:, 3));
                        
                        % 更新IMU点位置
                        for i = 1:5
                            set(imu_points(i), 'XData', spine_points(i, 1), ...
                                             'YData', spine_points(i, 2), ...
                                             'ZData', spine_points(i, 3));
                            
                            set(imu_labels(i), 'Position', spine_points(i, :));
                        end
                        
                        % 刷新图形
                        drawnow;
                    end
                end
            end
            
            % 短暂暂停
            pause(0.01);
            
        catch ME
            fprintf('错误: %s\n', ME.message);
            break;
        end
    end
    
    % 清理
    fclose(s);
    delete(s);
    clear s;
    fprintf('可视化已停止\n');
end

function spine_points = calculate_spine_points(imu_angles, segment_length)
    % 根据IMU角度计算脊柱各点位置
    spine_points = zeros(6, 3);
    current_pos = [0, 0, 0];
    current_orientation = eye(3);
    
    % 计算每个脊柱段的位置
    for i = 1:5
        % 获取当前IMU的欧拉角
        roll = imu_angles((i-1)*3 + 1);
        pitch = imu_angles((i-1)*3 + 2);
        yaw = imu_angles((i-1)*3 + 3);
        
        % 计算旋转矩阵
        R = euler_to_rotation_matrix(roll, pitch, yaw);
        
        % 更新当前方向
        current_orientation = R * current_orientation;
        
        % 计算下一段的方向向量 (Z轴方向)
        direction = current_orientation * [0; 0; 1];
        
        % 计算下一段的位置
        next_pos = current_pos + (direction * segment_length)';
        
        spine_points(i, :) = current_pos;
        current_pos = next_pos;
    end
    
    % 添加最后一个点
    spine_points(6, :) = current_pos;
end

function R = euler_to_rotation_matrix(roll, pitch, yaw)
    % 欧拉角转旋转矩阵
    % 转换为弧度
    roll = deg2rad(roll);
    pitch = deg2rad(pitch);
    yaw = deg2rad(yaw);
    
    % 旋转矩阵
    Rx = [1, 0, 0;
          0, cos(roll), -sin(roll);
          0, sin(roll), cos(roll)];
    
    Ry = [cos(pitch), 0, sin(pitch);
          0, 1, 0;
          -sin(pitch), 0, cos(pitch)];
    
    Rz = [cos(yaw), -sin(yaw), 0;
          sin(yaw), cos(yaw), 0;
          0, 0, 1];
    
    R = Rz * Ry * Rx;
end